### ISSUE TITLE
Executing an order along with stop loss and take profit on Binance

### ISSUE THREAD

**Issue by kooomix** [ccxt/ccxt#12924]
Hi,

I've been struggling lately with finding the right way to execute a limit/market order along with stop loss and take profit orders.

What I want to achieve is what is described in:
https://www.binance.com/en/support/faq/e1ee1738141c49718550fa9061be4bf3

So the end outcome will be 3 orders: primary limit/market order, a SL order and TP order.

Will be great if I can be referred to a code example to achieve this purpose.

Thanks a lot!
Eran

**Comment by samgermain** [comment]
[Here is a createOrder examples](https://github.com/ccxt/ccxt/blob/master/examples/js/create-order-handle-errors.js)
[Here is a createOrder example with a trigger price](https://github.com/ccxt/ccxt/blob/master/examples/js/binance-margin-stop-order.js)

For the stopPrice example, once the market price reaches the stopPrice specified by `params.stopPrice`, then an order will be created as normal, as if you were creating an order regularly at that moment

You can [read more about setting stop orders with CCXT here ](https://github.com/ccxt/ccxt/issues/12044)

**Comment by kooomix** [comment]
Thanks @samgermain for the reply.

I followed the example with triggered price and implemented as follows:



symbol = 'BTC/USDT'

type = 'limit'
side = 'buy'
amount = 0.001 
price = 40000

exchange_parameters = {'newClientOrderId': "myclientID111",
                        'type': 'spot',
                        'stopLossPrice': 39000,
                        'stopLossTimeInForce': "GTC",
                        'takeProfitPrice': 41000,
                        'takeProfitTimeInForce': "GTC",
                        'test': True
                          }

order = exchange.create_order(symbol, type, side, amount, price, params=exchange_parameters)

And getting the following error:
"BadRequest: binance {"code":-1104,"msg":"Not all sent parameters were read; read '11' parameter(s) but was sent '15'."}"

To recap - what i'm trying to achieve is to place a primary order of limit or market for spot, along with 2 secondaries orders - one for stop loss and one for take profit.

I'm using ccxt version '1.79.94'

What am I missing here?

Thanks.

**Comment by kroitor** [comment]
@kooomix Binance API does not allow you to set both a stop-loss (SL) and a take-profit (TP) on your position as you would do via the web interface. Instead, when working with the Binance API you place three separate orders, one order opening your position, and two stop orders for a stop-loss and a take-profit. 

See samgermain's links for examples on how to place stop orders above. 

Your code would look like this:

```
from pprint import pprint

order = exchange.create_order(symbol, 'limit', 'buy', amount, price)  # initial opening order
pprint(order)
sl = exchange.create_order(symbol, 'market', 'sell', amount, None, {'stopPrice': stopLossPrice})
pprint(sl)
tp = exchange.create_order(symbol, 'market', 'sell', amount, None, {'stopPrice': takeProfitPrice})
ppring(tp)
```

Also, check this related link from Binance first-hand:

- https://dev.binance.vision/t/how-to-implement-otoco-tp-sl-orders-using-api/1622

Hope that answers your question, let us know if not.

**Comment by kooomix** [comment]
Thanks @kroitor, I got it!

As for "market" order along with "stopPrice" i'm getting:
"InvalidOrder: binance stopPrice parameter is not allowed for BTC/USDT market orders"

Maybe it is possible to place stopPrice only for limit orders?

**Comment by kroitor** [comment]
> Thanks @kroitor, I got it!
> 
> As for "market" order along with "stopPrice" i'm getting: "InvalidOrder: binance stopPrice parameter is not allowed for BTC/USDT market orders"
> 
> Maybe it is possible to place stopPrice only for limit orders?

Yes, that is correct, pardon. You should use the `'limit'` order type and specify a limit price instead of `None`.

**Comment by kooomix** [comment]
I see, thanks for the quick respond!

**Comment by kooomix** [comment]
There is another issue - once I place a stop price order, the "amount" of the symbol gets "locked" until order is cancelled or filled, so if this was the total amount of the symbol, it is not possible to place another stop order until this amount is released...

Is there a way to buy X quantity for a certain symbol, and then place take profit and stop loss stop orders, each with X amount?

**Comment by kooomix** [comment]
Found the answer - seems like OCO does the trick:

https://github.com/ccxt/ccxt/issues/8241
